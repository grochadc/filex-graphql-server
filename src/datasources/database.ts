import pgPromise from "pg-promise";
import path from "path";
const { QueryFile } = pgPromise;
require("dotenv").config();

const initOptions = {
  connectionString: process.env.DATABASE_URL,
  ssl: { rejectUnauthorized: false }
};

console.log("Initializing database");
const pgp = pgPromise();
const db = pgp(initOptions);

export function sqlFile(file: string) {
  const fullPath = path.join(__dirname, file);
  const options = {
    minify: true
  };
  const qf = new QueryFile(fullPath, options);

  if (qf.error) {
    console.error(qf.error);
  }

  return qf;
}

export default db;
