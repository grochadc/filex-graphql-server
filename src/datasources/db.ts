type WorkshopId =
  | "conversation"
  | "toeflpreparation"
  | "basicreading"
  | "basicadvancedlistening"
  | "advancedreading"
  | "tutoring";

interface Workshop {
  id: WorkshopId;
  name: string;
  description: string;
  levels: number[];
  option_ids: string[];
}

interface Carrera {
  name: string;
}

const database: {
  carreras: Carrera[];
  workshops: Workshop[];
  options: any[];
  teachers: any[];
} = {
  carreras: [
    { name: "Academico" },
    { name: "Administrativo" },
    { name: "Abogado" },
    { name: "Administración de Negocios" },
    { name: "Agrobiotecnología" },
    { name: "Agronegocios" },
    { name: "Carrera en Enfermería (ENFE)" },
    {
      name:
        "Ciencia del Comportamiento con Orientación en Alimentación y Nutrición",
    },
    {
      name:
        "Ciencia del Comportamiento con orientación en Alimentación y Nutrición",
    },
    { name: "Cultura Física y Deportes" },
    { name: "Derecho" },
    { name: "Desarrollo Turístico Sustentable" },
    { name: "Enfermería" },
    { name: "Enfermería Semiescolarizada" },
    { name: "Estudios Socioterritoriales" },
    { name: "Ingeniería en Geofísica" },
    { name: "Ingeniería en Sistemas Biológicos" },
    { name: "Ingeniería en Telemática" },
    { name: "Letras Hispánicas" },
    { name: "Médico Cirujano y Partero" },
    { name: "Médico Veterinario y Zootecnista" },
    { name: "Negocios Internacionales" },
    { name: "Nivelación en Licenciatura en Enfermería" },
    { name: "Nutrición" },
    { name: "Periodismo" },
    { name: "Protección Civil y Emergencias" },
    { name: "Psicología" },
    { name: "Psicología con Orientación en Calidad de Vida y Salud" },
    { name: "Psicología con Orientación en Calidad de Vida y Salud" },
    { name: "Salud Pública" },
    { name: "Seguridad Laboral" },
    { name: "Tecnologías para el Aprendizaje" },
    { name: "Trabajo Social" },
  ],
  workshops: [
    {
      id: "conversation",
      name: "Conversation",
      description:
        "El alumno se expresa oralmente utilizando las funciones comunicativas y vocabulario adquirido, mejorar la fluidez, y perder el temor a hablar.",
      levels: [1, 2, 3, 4, 5, 6],
      option_ids: [
        "brendalunes",
        "erickalunes",
        "erickamartes",
        "sergiomartes",
        "jessicamartes",
        "carlosmiercoles",
        "sergiomiercoles",
        "gonzalojueves",
        "jessicajueves",
      ],
    },
    {
      id: "toeflpreparation",
      name: "TOEFL Preparation",
      description:
        "Ayuda a desarrollar habilidades para el examen de certificación, a partir del dominio ya adquirido del idioma. Duracion: 1 mes.",
      levels: [6],
      option_ids: [
        "carloslunes",
        "carlosmartes",
        "brendamiercoles",
        "brendajueves",
      ],
    },
    {
      id: "basicreading",
      name: "Basic Reading",
      description:
        "Lectura de textos e historias sencillas para trabajar con vocabulario y lectura de comprensión. Niveles 1-3",
      levels: [1, 2, 3],
      option_ids: ["erickamiercoles"],
    },
    {
      id: "advancedreading",
      name: "Advanced Reading",
      description:
        "Lectura de textos e historias sencillas para trabajar con vocabulario y lectura de comprensión. Niveles 1-3",
      levels: [4, 5, 6],
      option_ids: ["gisselmartes", "gisselmiercoles"],
    },
    {
      id: "basicadvancedlistening",
      name: "Basic/Advanced Listening",
      description:
        "Ayuda a desarrollar la habilidad de comprensión de escucha por medio de audios. Basic: Niveles 1-3; Advanced: Niveles 4-6",
      levels: [1, 2, 3, 4, 5, 6],
      option_ids: ["zulletlunes", "zulletmartes"],
    },
    {
      id: "tutoring",
      name: "Tutoring",
      description:
        "Es un proceso de acompañamiento, aclarando dudas. El alumno deberá saber exactamente el tema a repasar.",
      levels: [1, 2, 3, 4, 5, 6],
      option_ids: ["sergiolunes", "jessicamiercoles", "gisseljueves"],
    },
  ],
  options: [
    {
      id: "carloslunes",
      teacher_id: "carlos",
      time: "12:00 - 13:00",
      day: "lunes",
      workshop_id: "toeflpreparation",
      url: "https://classroom.google.com/c/MTU0NzM2MzkzOTY3?cjc=txkvhcl",
    },
    {
      id: "brendalunes",
      teacher_id: "brenda",
      time: "12:00 - 13:00",
      day: "lunes",
      workshop_id: "conversation",
      url:
        "https://us04web.zoom.us/j/77370162343?pwd=VFpCS2xqejF1blJvY2t4OWk1c3RtQT09",
    },
    {
      id: "zulletlunes",
      teacher_id: "zullet",
      time: "12:00 - 13:00",
      day: "lunes",
      workshop_id: "basicadvancedlistening",
      url: "https://classroom.google.com/c/MjI0NzkzNjM4NjU2?cjc=3pqkhy4",
    },
    {
      id: "erickalunes",
      teacher_id: "ericka",
      time: "13:00 - 14:00",
      day: "lunes",
      workshop_id: "conversation",
      url: "https://udg-mx.zoom.us/u/kdPy130e6a",
    },
    {
      id: "sergiolunes",
      teacher_id: "sergio",
      time: "14:00 - 15:00",
      day: "lunes",
      workshop_id: "tutoring",
      url: "https://meet.google.com/yvi-qirh-gbs",
    },
    {
      id: "gisselmartes",
      teacher_id: "gissel",
      time: "11:00 - 12:00",
      day: "martes",
      workshop_id: "advancedreading",
      url: "https://classroom.google.com/u/1/c/MjcwMzg3ODY3MjM4",
    },
    {
      id: "carlosmartes",
      teacher_id: "carlos",
      time: "12:00 - 13:00",
      day: "martes",
      workshop_id: "toeflpreparation",
      url: "https://classroom.google.com/c/MTU0NzM2MzkzOTY3?cjc=txkvhcl",
    },
    {
      id: "zulletmartes",
      teacher_id: "zullet",
      time: "12:00 - 13:00",
      day: "martes",
      workshop_id: "basicadvancedlistening",
      url: "https://classroom.google.com/c/MjI0NzkzNjM4NjU2?cjc=3pqkhy4",
    },
    {
      id: "erickamartes",
      teacher_id: "ericka",
      time: "13:00 - 14:00",
      day: "martes",
      workshop_id: "conversation",
      url: "https://udg-mx.zoom.us/u/kdPy130e6a",
    },
    {
      id: "sergiomartes",
      teacher_id: "sergio",
      time: "14:00 - 15:00",
      day: "martes",
      workshop_id: "conversation",
      url: "https://meet.google.com/yvi-qirh-gbs",
    },
    {
      id: "jessicamartes",
      teacher_id: "jessica",
      time: "16:00 - 17:00",
      day: "martes",
      workshop_id: "conversation",
      url:
        "https://us04web.zoom.us/j/5461493724?pwd=djBZZFVSUWJuNG5kSVhkSGVUWnA1QT09",
    },
    {
      id: "gisselmiercoles",
      teacher_id: "gissel",
      time: "11:00 - 12:00",
      day: "miercoles",
      workshop_id: "advancedreading",
      url: "https://classroom.google.com/u/1/c/MjcwMzg3ODY3MjM4",
    },
    {
      id: "brendamiercoles",
      teacher_id: "brenda",
      time: "11:00 - 12:00",
      day: "miercole",
      workshop_id: "toeflpreparation",
      url: "https://classroom.google.com/c/MjczODExMzE2NzEw?cjc=puxzawj",
    },
    {
      id: "jessicamiercoles",
      teacher_id: "jessica",
      time: "11:00 - 12:00",
      day: "miercoles",
      workshop_id: "tutoring",
      url:
        "https://us04web.zoom.us/j/5461493724?pwd=djBZZFVSUWJuNG5kSVhkSGVUWnA1QT09",
    },
    {
      id: "carlosmiercoles",
      teacher_id: "carlos",
      time: "12:00 - 13:00",
      day: "miercoles",
      workshop_id: "conversation",
      url:
        "https://us02web.zoom.us/j/86845226720?pwd=dTVmbllvaXpDLzJneDV1TEVVVU83Zz09",
    },
    {
      id: "erickamiercoles",
      teacher_id: "ericka",
      time: "13:00 - 14:00",
      day: "miercoles",
      workshop_id: "basicreading",
      url: "classroom",
      zoom_id: "4ty4g44",
    },
    {
      id: "sergiomiercoles",
      teacher_id: "sergio",
      time: "14:00 - 15:00",
      day: "miercoles",
      workshop_id: "conversation",
      url: "https://meet.google.com/yvi-qirh-gbs",
    },
    {
      id: "brendajueves",
      teacher_id: "brenda",
      time: "11:00 - 12:00",
      day: "jueves",
      workshop_id: "toeflpreparation",
      url: "https://classroom.google.com/c/MjczODExMzE2NzEw?cjc=puxzawj",
    },
    {
      id: "gonzalojueves",
      teacher_id: "gonzalo",
      time: "15:00 - 16:00",
      day: "jueves",
      workshop_id: "conversation",
      url: "http://meet.google.com/lookup/gonzaloconversation",
    },
    {
      id: "gisseljueves",
      teacher_id: "gissel",
      time: "16:00 - 17:00",
      day: "jueves",
      workshop_id: "tutoring",
      url: "https://meet.google.com/nrb-bdav-hru",
    },
    {
      id: "jessicajueves",
      teacher_id: "jessica",
      time: "16:00 - 17:00",
      day: "jueves",
      workshop_id: "conversation",
      url:
        "https://us04web.zoom.us/j/5461493724?pwd=djBZZFVSUWJuNG5kSVhkSGVUWnA1QT09",
    },
  ],
  teachers: [
    {
      id: "jessica",
      name: "Jessica",
      options: ["jessicamartes", "jessicamiercoles", "jessicajueves"],
    },
    {
      id: "ericka",
      name: "Ericka",
      options: ["erickalunes", "erickamartes", "erickamiercoles"],
    },
    {
      id: "gonzalo",
      name: "Gonzalo",
      options: ["gonzalojueves"],
    },
    {
      id: "sergio",
      name: "Sergio",
      options: ["sergiolunes", "sergiomartes", "sergiomiercoles"],
    },
    {
      id: "gissel",
      name: "Gissel",
      options: ["gisselmartes", "gisseljueves", "gisselmiercoles"],
    },
    {
      id: "carlos",
      name: "Carlos",
      options: ["carloslunes", "carlosmartes", "carlosmiercoles"],
    },
    {
      id: "zullet",
      name: "Zullet",
      options: ["zulletlunes", "zulletmartes"],
    },
    {
      id: "brenda",
      name: "Brenda",
      options: ["brendalunes", "brendamiercoles", "brendajueves"],
    },
  ],
};

export default database;
