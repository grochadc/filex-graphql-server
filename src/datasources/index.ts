import { FirebaseAPI } from "./myREST";
import { RegistroAPI } from "./RegistroAPI";
import { ExamAPI } from "./ExamAPI";
import { WorkshopsAPI } from "./WorkshopsAPI";
import { StudentsAPI } from "./StudentsAPI";

export { RegistroAPI, FirebaseAPI, ExamAPI, WorkshopsAPI, StudentsAPI };
