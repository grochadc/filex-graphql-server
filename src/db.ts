type WorkshopId =
  | "conversation"
  | "toeflpreparation"
  | "basicreading"
  | "basicadvancedlistening"
  | "tutoring";

interface Workshop {
  id: WorkshopId;
  name: string;
  description: string;
}

interface WorkshopOption {
  id: string;
  teacher: string;
  time: string;
  day: "lunes" | "martes" | "miercoles" | "jueves";
  workshop_id: WorkshopId;
}

export interface Applicant {
  id: string;
  code: string;
  name: string;
  workshop_id: WorkshopId;
  option_id: string;
}

const db: {
  workshops: Workshop[];
  options: WorkshopOption[];
  applicants: Applicant[];
} = {
  workshops: [
    {
      id: "conversation",
      name: "Conversation",
      description:
        "El alumno se expresa oralmente utilizando las funciones comunicativas y vocabulario adquirido, mejorar la fluidez, y perder el temor a hablar.",
    },
    {
      id: "toeflpreparation",
      name: "TOEFL Preparation",
      description:
        "Ayuda a desarrollar habilidades para el examen de certificación, a partir del dominio ya adquirido del idioma. Duracion: 1 mes.",
    },
    {
      id: "basicreading",
      name: "Basic Reading",
      description:
        "Lectura de textos e historias sencillas para trabajar con vocabulario y lectura de comprensión. Niveles 1-3",
    },
    {
      id: "basicadvancedlistening",
      name: "Basic/Advanced Listening",
      description:
        "Ayuda a desarrollar la habilidad de comprensión de escucha por medio de audios. Basic: Niveles 1-3; Advanced: Niveles 4-6",
    },
    {
      id: "tutoring",
      name: "Tutoring",
      description:
        "Es un proceso de acompañamiento, aclarando dudas. El alumno deberá saber exactamente el tema a repasar.",
    },
  ],
  options: [
    {
      id: "alondralunes",
      teacher: "alondra",
      time: "13:00 - 14:00",
      day: "lunes",
      workshop_id: "conversation",
    },
    {
      id: "gonzalolunes",
      teacher: "gonzalo",
      time: "14:00 - 15:00",
      day: "lunes",
      workshop_id: "conversation",
    },
    {
      id: "alondramartes",
      teacher: "alondra",
      time: "13:00 - 14:00",
      day: "martes",
      workshop_id: "conversation",
    },
    {
      id: "sergiomartes",
      teacher: "sergio",
      time: "14:00 - 15:00",
      day: "martes",
      workshop_id: "conversation",
    },
    {
      id: "gisselmartes",
      teacher: "gissel",
      time: "16:00 - 17:00",
      day: "martes",
      workshop_id: "conversation",
    },
    {
      id: "carlosmiercoles",
      teacher: "carlos",
      time: "12:00 - 13:00",
      day: "miercoles",
      workshop_id: "conversation",
    },
    {
      id: "sergiomiercoles",
      teacher: "sergio",
      time: "14:00 - 15:00",
      day: "miercoles",
      workshop_id: "conversation",
    },
    {
      id: "gisselmiercoles",
      teacher: "gissel",
      time: "16:00 - 17:00",
      day: "miercoles",
      workshop_id: "conversation",
    },
    {
      id: "gonzalojueves",
      teacher: "gonzalo",
      time: "14:00 - 15:00",
      day: "jueves",
      workshop_id: "conversation",
    },
    {
      id: "zulletjueves",
      teacher: "zullet",
      time: "13:00 - 14:00",
      day: "jueves",
      workshop_id: "conversation",
    },
    {
      id: "carloslunes",
      teacher: "carlos",
      time: "12:00 - 13:00",
      day: "lunes",
      workshop_id: "toeflpreparation",
    },
    {
      id: "carlosmartes",
      teacher: "carlos",
      time: "12:00 - 13:00",
      day: "martes",
      workshop_id: "toeflpreparation",
    },
    {
      id: "jiselainemiercoles",
      teacher: "jiselaine",
      time: "11:00 - 12:00",
      day: "miercoles",
      workshop_id: "toeflpreparation",
    },
    {
      id: "jiselainejueves",
      teacher: "jiselaine",
      time: "11:00 - 12:00",
      day: "jueves",
      workshop_id: "toeflpreparation",
    },
    {
      id: "alondramiercoles",
      teacher: "alondra",
      time: "13:00 - 14:00",
      day: "miercoles",
      workshop_id: "basicreading",
    },
    {
      id: "zulletlunes",
      teacher: "zullet",
      time: "12:00 - 13:00",
      day: "lunes",
      workshop_id: "basicadvancedlistening",
    },
    {
      id: "zulletmartes",
      teacher: "zullet",
      time: "12:00 - 13:00",
      day: "martes",
      workshop_id: "basicadvancedlistening",
    },
    {
      id: "sergiolunes",
      teacher: "sergio",
      time: "14:00 - 15:00",
      day: "lunes",
      workshop_id: "tutoring",
    },
    {
      id: "gissellunes",
      teacher: "gissel",
      time: "16:00 - 17:00",
      day: "lunes",
      workshop_id: "tutoring",
    },
    {
      id: "zulletmiercoles",
      teacher: "zullet",
      time: "11:00 - 12:00",
      day: "miercoles",
      workshop_id: "tutoring",
    },
  ],
  applicants: [
    {
      id: "4kmlkb",
      code: "52841",
      name: "Juan Gabriel",
      workshop_id: "conversation",
      option_id: "alondralunes",
    },
    {
      id: "qaveis",
      code: "12345",
      name: "Pedro Paramo",
      workshop_id: "conversation",
      option_id: "alondramartes",
    },
    {
      id: "rww626",
      code: "54321",
      name: "Lenin",
      workshop_id: "conversation",
      option_id: "alondralunes",
    },
  ],
};

export default db;
