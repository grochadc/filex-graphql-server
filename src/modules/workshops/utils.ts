export function sortWorkshops(options) {
  const dict = {
    Lunes: 1,
    Martes: 2,
    Miercoles: 3,
    Jueves: 4
  };

  const byDay = (a, b) => {
    const left = dict[a.day];
    const right = dict[b.day];
    return left - right;
  };
  const byTime = (a, b) => {
    return Number(a.time.substring(0, 2)) - Number(b.time.substring(0, 2));
  };
  const sortedByDay = options.slice().sort(byDay);
  const sortedByTime = sortedByDay.slice().sort(byTime);
  return sortedByTime;
}
