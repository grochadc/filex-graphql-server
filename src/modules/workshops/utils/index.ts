import {
  Student as PrismaStudent,
  Applicant as PrismaApplicant,
  Group as PrismaGroup,
} from "@prisma/client";
import { Student as GraphStudent } from "generated/graphql";
export function sortWorkshops(options) {
  const dict = {
    Lunes: 1,
    Martes: 2,
    Miercoles: 3,
    Jueves: 4,
  };

  const byDay = (a, b) => {
    const left = dict[a.day];
    const right = dict[b.day];
    return left - right;
  };
  const byTime = (a, b) => {
    return Number(a.time.substring(0, 2)) - Number(b.time.substring(0, 2));
  };
  const sortedByDay = options.slice().sort(byDay);
  const sortedByTime = sortedByDay.slice().sort(byTime);
  return sortedByTime;
}

export function sortStudents(students: { nombre: string }[]) {
  const byName = (
    left: { nombre: string },
    right: { nombre: string }
  ): number => {
    return left.nombre.localeCompare(right.nombre);
  };
  return students.sort(byName);
}

type StudentBeforeUnwind = PrismaStudent & {
  applicant: PrismaApplicant;
  groupObject: PrismaGroup;
};

type PropType<TObj, TProp extends keyof TObj> = TObj[TProp];

type MappedProps = 
  Record<"ciclo", PropType<PrismaStudent, "ciclo_actual">> &
  Record<"grupo", PropType<PrismaGroup, "name">>;



type StudentAfterUnwind = PrismaStudent & PrismaApplicant & MappedProps;

export function unwindPrismaStudent(
  student: StudentBeforeUnwind
): StudentAfterUnwind {
  return {
    ...student,
    ...student.applicant,
    ciclo: student.ciclo_actual,
    grupo: student.groupObject.name,
  };
}


