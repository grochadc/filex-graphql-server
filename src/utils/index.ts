const { groupBy } = require("ramda");
const { Option } = require("../datasources/db");
type MatcherFn = (iteration: any, actual: any) => boolean;
import { MeetLink } from "../types/index";

export const getById = (
  obj: any,
  key: string,
  id: string,
  callback: () => void
) => {
  if (callback) callback();
  const data = obj[key].filter((item) => {
    return item.id === id;
  });
  if (data.length === 0) throw new Error(`Couldn't find ${id} in ${key}`);
  return data[0];
};

export const getByIds = (arr: any[], ids: string[]) => {
  return arr.filter((item) => ids.indexOf(item.id) > -1);
};

export const generateId = () => {
  let str = Math.random().toString(36).substring(7);
  return str;
};

export const createMatrix = (arr) => {
  return arr.map((item) => Object.values(item));
};

type WorkshopId =
  | "conversation"
  | "toeflpreparation"
  | "basicreading"
  | "basicadvancedlistening"
  | "advancedreading"
  | "tutoring";

interface WorkshopOption {
  id: WorkshopId;
  name: string;
  description: string;
  levels: number[];
}
export const mapOptionIds = (options: Option[], option_ids: string[]) =>
  option_ids.map((id) => options.filter((option) => option.id === id)[0]);

interface Option {
  id: string;
  teacher_id: string;
  time: string;
  day: string;
  workshop_id: string;
  url: string;
}

interface Teacher {
  id: string;
  name: string;
  options: string[];
}

export const mapOptionsTeacher = (options: Option[], teachers: Teacher[]) =>
  options.map(
    (option) =>
      teachers.filter((teacher) => teacher.id === option.teacher_id)[0]
  );

export function getIndexToModify(
  item: any,
  array: any[],
  matcher: MatcherFn = (iteration, actual) => iteration === actual
): number {
  const reducedIndex = array.reduce(
    (acc, curr, index) => (matcher(curr, item) ? index : acc + 0),
    -1
  );
  return reducedIndex > -1 ? reducedIndex : array.length;
}

export function addIDsToLinks(links: MeetLink[]): MeetLink[] {
  const anyUndefined = links.some((link) => link.id === undefined);
  if (anyUndefined) {
    return links.map((link) =>
      link.id === undefined ? { ...link, id: generateId() } : link
    );
  }
  return links;
}
