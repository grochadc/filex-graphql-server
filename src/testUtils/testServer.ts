import {
  createTestClient,
  ApolloServerTestClient
} from "apollo-server-testing";
import { ApolloServer } from "apollo-server";

export default function testServer(
  dataSources?: () => any,
  context?: () => any
): ApolloServerTestClient {
  return createTestClient(
    new ApolloServer({
      modules: [
        require("../modules/placement_exam"),
        require("../modules/placement_settings"),
        require("../modules/registro"),
        require("../modules/workshops"),
        require("../modules/test_questions")
      ],
      dataSources,
      context
    })
  );
}
