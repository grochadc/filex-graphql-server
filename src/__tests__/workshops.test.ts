import testServer from "../testUtils/testServer";
import { WorkshopsAPI, StudentsAPI } from "../datasources";
import {
  GET_RESERVATIONS,
  GET_SELECTION_INFO,
  SET_RESERVATION,
} from "./__data__/queries";
import {
  reservations,
  options,
  student,
  workshopOption,
} from "./__data__/mocks";
import db from "../datasources/db";
import MockDate from "mockdate";
import * as utils from "../utils";

const { workshops } = db;

test("serves reservations and teacher info", async () => {
  const workshopsAPI = new WorkshopsAPI();
  workshopsAPI.getReservations = jest.fn((teacher_id: string) => {
    return Promise.resolve(reservations);
  });
  //workshopsAPI.getOptionsByTeacherId = jest.fn((teacher_id: string) => options);

  const { query } = testServer(() => ({ workshopsAPI }));
  const res = await query({
    query: GET_RESERVATIONS,
    variables: { teacher: "sergio" },
  });
  if (res.errors) console.log(res.errors);
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();

  expect(workshopsAPI.getReservations).toHaveBeenCalledWith("sergio");
  expect(workshopsAPI.getOptionsByTeacherId).toHaveBeenCalledWith("sergio");
});

test("serves info for selection page", async () => {
  const studentsAPI = new StudentsAPI();
  const workshopsAPI = new WorkshopsAPI();

  studentsAPI.getStudent = jest.fn(() => Promise.resolve(student));
  workshopsAPI.getWorkshops = jest.fn(() => workshops);

  const { query } = testServer(() => ({ studentsAPI, workshopsAPI }));
  const res = await query({
    query: GET_SELECTION_INFO,
    variables: { code: "1234567890" },
  });
  if (res.errors) console.log(res.errors);
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();

  expect(studentsAPI.getStudent).toHaveBeenCalledWith("1234567890");
  expect(workshopsAPI.getWorkshops).toHaveBeenCalled();
});

test("saves a reservation", async () => {
  const studentsAPI = new StudentsAPI();
  const workshopsAPI = new WorkshopsAPI();

  studentsAPI.getStudent = jest.fn(() => Promise.resolve(student));
  workshopsAPI.getOptionById = jest.fn((option_id: string) => workshopOption);
  const spy = jest.spyOn(utils, "generateId").mockReturnValueOnce("w924pj");
  MockDate.set("2000-11-22");

  const { mutate } = testServer(() => ({ studentsAPI, workshopsAPI }));

  const res = await mutate({
    query: SET_RESERVATION,
    variables: { codigo: "1234567890", option_id: "gonzalojueves" },
  });

  if (res.errors) console.log(res.errors);
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();

  expect(spy).toHaveBeenCalled();
  expect(studentsAPI.getStudent).toHaveBeenCalledWith("1234567890");
  expect(workshopsAPI.getOptionById).toHaveBeenCalledWith("gonzalojueves");
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();
});
