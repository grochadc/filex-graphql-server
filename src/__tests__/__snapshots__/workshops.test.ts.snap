// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets all workshops 1`] = `
Object {
  "workshops": Array [
    Object {
      "description": "This is the conversation workshop description.",
      "levels": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
      ],
      "name": "Conversation",
      "options": Array [
        Object {
          "available": true,
          "day": "Lunes",
          "id": "id1",
          "teacher_name": "Fulanito",
          "time": "1300 - 14:00",
          "url": "http://url1",
          "workshop_name": "Conversation",
          "zoom_id": null,
        },
        Object {
          "available": true,
          "day": "Martes",
          "id": "id2",
          "teacher_name": "Fulanito",
          "time": "13:00 - 14:00",
          "url": "http://url2",
          "workshop_name": "Conversation",
          "zoom_id": null,
        },
      ],
    },
  ],
}
`;

exports[`makes a reservation 1`] = `
Object {
  "apellido_materno": "Ocasio",
  "apellido_paterno": "Martinez",
  "codigo": "1234567890",
  "grupo": "E4-1",
  "nivel": "4",
  "nombre": "Benito Antonio",
  "option_id": "id1",
  "option_name": "why? option_name from makeReservation dataSource",
  "workshop_id": "workshop_id1",
  "workshop_name": "Conversation",
}
`;

exports[`student data already registered 1`] = `
Object {
  "student": Object {
    "nombre": "Juan",
    "reservation": Object {
      "day": "Lunes",
      "teacher_name": "Fulanito",
      "time": "1300 - 14:00",
      "url": "http://url1",
      "workshop_name": "Conversation",
    },
  },
}
`;

exports[`student data normal not registered 1`] = `
Object {
  "student": Object {
    "nombre": "Benito Antonio",
    "reservation": null,
  },
}
`;

exports[`teacher dashboard no reservations 1`] = `
Object {
  "teacher": Object {
    "id": "teacher_id2",
    "name": "Manganito",
    "options": Array [
      Object {
        "day": "Lunes",
        "id": "id1",
        "reservations": null,
        "time": "1300 - 14:00",
        "url": "http://url1",
      },
    ],
  },
}
`;

exports[`teacher dashboard normal 1`] = `
Object {
  "teacher": Object {
    "id": "teacher_id1",
    "name": "Fulanito",
    "options": Array [
      Object {
        "day": "Lunes",
        "id": "id1",
        "reservations": Array [
          Object {
            "apellido_materno": "Ocasio",
            "apellido_paterno": "Martinez",
            "codigo": "1234567890",
            "grupo": "E4-1",
            "nivel": "4",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "1300 - 14:00",
        "url": "http://url1",
      },
      Object {
        "day": "Martes",
        "id": "id2",
        "reservations": null,
        "time": "13:00 - 14:00",
        "url": "http://url2",
      },
    ],
  },
}
`;
